### ISSUE TITLE
No simplification for union of symbolic Interval

### ISSUE THREAD

**Issue by TApplencourt** [sympy/sympy#12461]
I have two Sympy interval:
```(Interval(1, m), Interval(1, m+1)) ``` 

To make the union of those interval you can do:
```
m = symbols('m', integer=True)
Union(Interval(1, m), Interval(1, m+1))
>> [1, m] U [1, m + 1]
```
But now I and to simplify this expression.  I know that `m>=1` so the union should be  `[1, m+1]`.  Unfortunately It not possible (to my knowledge) to use this constraint on `m`.

##### Nota Bene
A quick  and naive mock-up of the desired result to make things more concrete:
```
m = symbols('m', integer=True)
u = Union(Interval(1, m), Interval(1, m+1)
solveset(u,m, [m>=1]) # Proposition
>> [1, m+1]
```

**Comment by asmeurer** [comment]
I think the union should be [1, m+1] regardless whether m is greater than 1. If m is less than 1 then [1, m] is empty, so the union of the empty set with another set is just the other set.

**Comment by TApplencourt** [comment]
Good catch! I totally agree with you. The simpy result was not false but
can be simplify as you explained.


So in think we have now to aspect. First it will be useful if we can
simplify the result of the Union and  secondly if we can restrict the
domain of the symbol to solve the union in much complicated case.

Indeed, my example was over simplified. Let say that i want to compute
```
u = [A,m] U [m,B]
```
where, without loosing generally, `A < B` with `A`,`B` are some arbitrary
constant.

No we have those 3 cases for `u` (if I'm correct):
- if `A <m < B` then  `u = [A, B] `
- if `m < A` then `u = [m, B]`
- if `m > B` then `u = [A, m]`

With the restrictions in all cases that `[a,b] = empy set` if `a>b`.

**Comment by asmeurer** [comment]
I believe that is correct. It's unfortunate, because with more than two intervals the number of inequalities explodes combinatorially, unless the end points are numeric or have an obvious relationship (like m and m+1).  So at best you'd need to specify the inequalities for it to simplify, which feels redundant. It would seem to be more useful to always be able to assume that [a, b] means a <= b (which can still be empty if a = b). I think we shouldn't necessarily do this in general, but there should be an option to. 

At any rate, it should be possible to simplify the original union without any advanced logic.

**Comment by zimoun** [comment]
Assuming strict order about symbols, such that: `a<b` leads to combinatorial number of inequalities, but somehow you **have to** provide one relation order about your symbolic objects coming from your reference set (integer, real, other). Otherwise you cannot reduce.

Assuming strict order, the reduction is simple logic. Combinatorial but simple :-)
Even there is open intervals or infinity which need carefulness.

However, if you authorize non-strict order, such `a<=b`, then the reduction is becoming much more complicated, from my attempts.

Note that `[a,b]` usually means any integer `x` such `x>a` and `x<b`.
Therefore, if `a=b` then `[a,b]={}` is the empty set.

Otherwise, if `[a,b]` means any integer `x` such `x>=a` and `x<=b`.
then `a=b` implies that `[a,b]={a}` is the singleton.

Anyway! :-)

What *should* be nice: provide the relation order about the used symbols through Assumptions module.

In the flavor:
```python
A, m, B = symbols("A m B")

let = Q.positive(A) & Q.positive(m) & Q.positive(B)
rel = Q.is_true(A < m) & Q.is_true(m < B)
facts = let & rel

u = Union(Interval(A, m), Interval(m, B), facts)
# returns [A, B]
```

**Comment by asmeurer** [comment]
You can represent the inequalities with the assumptions, but it's annoying to type `A < m` and `m < B` in addition to `Interval(A, m)` and `Interval(m, B)`. It feels redundant, because if you create an interval, you probably meant for it to be nonempty. This isn't true for intervals that are created programmatically or generated as a result of some other set operation, however.

**Comment by zimoun** [comment]
Let me clarify my point.

Currently, `Interval(a, b)` means any `x` in real such that `a<=x<=b`.
If `a>b` then it is the `EmptySet`.
If `a=b` then it is the singleton `{a}`.

If you consider the simple example `u=[A,m] U [m, B]`, then by definition of `Interval`, the 3 cases read as above. I mean, obtain the right reduction is obvious here. And the tools are already there.

Now, let consider examples to point out possible extensions and/or motivations to improve the Assumption module.

Let `u=[a, b] U [c, d]`. Then simple logic is straightforward:
 - `b<c` or `d<a` implies that `u` is already reduced. 
Note that if `a=b` or `c=d` is provided then more reduction is possible.
 -  `c<b` and ( `a<c` or `b<d` ) implies that `u=[a,d]`
 - etc.
 - and `oo` needs special treatment. Open interval `(a,b)` too.

It is not too difficult to write an overlay to specify the relation order between `a, b, c, d`. 

However, the question is: **should be nice to use Assumption module ?**
Especially if you imagine more complicated example such that `u=[A+m,3*m+1] U [m-1,B-m]`.

In the flavor:
```python
A, m, B = symbols("A m B", integer=True)

let = Q.positive(A) & Q.positive(m) & Q.positive(B)
rel = Q.is_true(A < m) & Q.is_true(3*m-1 < B)
facts = let & rel

u = Union(Interval(A+m, 3*m+1), Interval(m-1, B-m), facts)
# returns [A+m, B-m]
```

---
I use the kind of last example when I evaluate bounds to pass to `integrate` for simple functions (polynomials, etc.).

**Comment by oscarbenjamin** [comment]
This is still the case on current master:
```julia
In [43]: m = symbols('m', integer=True) 
    ...: Union(Interval(1, m), Interval(1, m+1))                                                                                                              
Out[43]: [1, m] ∪ [1, m + 1]

In [44]: simplify(_)                                                                                                                                          
Out[44]: [1, m] ∪ [1, m + 1]
```
